sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/storage/GcsMatchIT.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.storage;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import com.google.common.io.ByteStreams;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.PipelineResult.State;
+import org.apache.beam.sdk.extensions.gcp.options.GcsOptions;
+import org.apache.beam.sdk.extensions.gcp.util.gcsfs.GcsPath;
+import org.apache.beam.sdk.io.FileIO;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestPipelineOptions;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.Watch;
+import org.apache.beam.sdk.util.MimeTypes;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.joda.time.Duration;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class GcsMatchIT {
+  /** Integration test for TextIO.MatchAll watching for file updates in gcs filesystem */
+  @Test
+  public void testGcsMatchContinuously() throws InterruptedException {
+    TestPipelineOptions options =
+        TestPipeline.testingPipelineOptions().as(TestPipelineOptions.class);
+    assertNotNull(options.getTempRoot());
+    options.setTempLocation(options.getTempRoot() + "/testGcsMatchContinuouslyTest");
+    GcsOptions gcsOptions = options.as(GcsOptions.class);
+    String dstFolderName =
+        gcsOptions.getGcpTempLocation()
+            + String.format(
+                "/GcsMatchIT-%tF-%<tH-%<tM-%<tS-%<tL.testGcsMatchContinuously.copy/", new Date());
+    final GcsPath watchPath = GcsPath.fromUri(dstFolderName);
+
+    Pipeline p = Pipeline.create(options);
+
+    PCollection<Metadata> matchAllUpdatedMetadata =
+        p.apply("create for matchAll updated files", Create.of(watchPath.resolve("*").toString()))
+            .apply(
+                "matchAll updated",
+                FileIO.matchAll()
+                    .continuously(
+                        Duration.millis(250),
+                        Watch.Growth.afterTimeSinceNewOutput(Duration.standardSeconds(3)),
+                        true));
+
+    // Copy the files to the "watch" directory;
+    Thread writer =
+        new Thread(
+            () -> {
+              try {
+                Thread.sleep(1000);
```

Comment:
It makes sense that we would need to sleep in a test to validate this functionality, but sleeping in tests is not usually a good practice. @chamikaramj  or @pabloem is there a better pattern to use here?

Reply:
This test basically duplicates `org.apache.beam.sdk.io.testMatchWatchForNewFiles` and I am aware of that adding sleep makes the test time consuming. Also curious to see if there is a good idea